compaction/compaction.cc

```diff
@@ -729,6 +731,11 @@ class compaction {
 
         _ms_metadata.min_timestamp = timestamp_tracker.min();
         _ms_metadata.max_timestamp = timestamp_tracker.max();
+
+        size_t max_overlapping = std::ranges::max(token_map | boost::adaptors::transformed([] (auto& interval) {
+            return interval.second.size();
+        }));
+        _incremental_compaction_worthwhile = max_overlapping < _sstables.size();
```

Comment:
I think this generates some false negatives. If we have one run with two sstables S1, S2, and another run with one sstable S3, we get four elements in token_map: one for S1 only, one for S1+S3, one for S2+S3, one for S2. But it should definitely be enabled for incremental compaction.

Reply:
Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

Reply:
then (max_overlapping == 2) < (_sstables.size() == 3). incremental is worth being enabled if we detect that at any point in time there will be less sstables opened than there are in total, and that's what I implemented.

Reply:
> Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

That's a good counter example. So I can change it to enable incremental if there's any non-overlapping sstable.

Reply:
Actually, that's wrong, as there could be overlapping among all intervals, but it's still worth incremental.

Reply:
> Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

Wait, that would still be correct. 200 < 201, so incremental would be enabled, and that's correct because the disjoint sstable (or the overlapping ssts; depending on their token order) can be released earlier. False positive is ok (just some additional logic seeing nothing can be released), false negative is not.

Reply:
Then why not always be incremental? Is there a downside?

I guess generation of the gc sstable is a downside.

But I don't trust the calculation. How do we know it doesn't generate more false positives, or false negatives?

Reply:
We still only enable incremental compaction for ICS and LCS. We check max allowed size for output fragments too. So STCS and TWCS are unaffected by this change.

Effect of today's _contains_multi_fragment_runs is that any level above 0 will enable incremental compaction.

On "level 0", max overlapping == # of input runs.

Even on levels greater than 0, where memtables are way smaller than 0, max overlapping is still == # of runs.

That's because compaction output a single-fragment run if the output size is < 1G.

With this new calculation, that's preserved.

max overlapping will only be equal to # of input sstables on levels that contain single-fragment runs. And in practice, level 0 sstables all overlap one another at some token T.
In other words, "level 0" compaction's max overlapping will equal the number of input sstables.
level 1 and above, containing multi-fragment runs, will have less overlapping than the number of input sstables.

So the semantics is still the same, but we no longer rely on run id to determine activation of incremental.

Reply:
ok, makes sense.

So what does the test check? that the maximum overlap is less than the number of sstables? It's a really low bar.

Reply:
So why not it unconditionally for LCS/ICS?

Reply:
single large table == single table taking more than 50% of space.

Reply:
I mean, we're doing a lot of work with the interval_map. If it will often say that incremental is fine, why not say it every time? I don't want to look for stalls if there are lots of input sstables.

AFAICT it will say true if there is any disjoint sstable. Maybe we can find a simpler test for it?

Reply:
An ugly and lazy alternative is to keep contains_multi_fragment_runs, expose a force_incremental_compaction, which off-strategy sets to true. But you're right. I think we should just unconditionally enable incremental for ics / lcs. On the regular path, the cost of enabling incremental on level 0 is low to almost none.